platform/mds/mds-web/src/main/java/org/motechproject/mds/web/service/impl/InstanceServiceImpl.java

```diff
@@ -952,25 +953,32 @@ private Object parseValueForDisplay(Object value, MetadataDto relatedFieldMetada
             //Using ZonedDateTime for retrieving timezone
             ZonedDateTime zonedDateTime = ZonedDateTime.of((LocalDateTime) parsedValue, ZoneOffset.systemDefault());
             parsedValue = java.time.format.DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm xxxx").format(zonedDateTime);
-        } else if (relatedFieldMetadata != null) {
-            parsedValue = removeCircularRelations(parsedValue, relatedFieldMetadata.getValue());
+        } else if (relatedClassMetadata != null) {
+            // We do not want to return the whole chain of relationships for UI display, but just the first level.
+            // Fetching whole relationship tree may cause trouble when serializing
+            parsedValue = breakDeepRelationChainForDisplay(parsedValue, relatedClassMetadata.getValue());
         }
 
         return parsedValue;
     }
 
-    private Object removeCircularRelations(Object object, String relatedField) {
-        // we must also handle a field that is a collection
-        // because of this we handle regular fields as single objects collection here
-        Collection objectsCollection = (object instanceof Collection) ? (Collection) object : Arrays.asList(object);
+    private Object breakDeepRelationChainForDisplay(Object value, String relatedClassName) {
```

Comment:
That being said, the best way to handle this would be to move the method to a separate helper component. That way you can use the Transactional annotation with propagation=NEVER on this method. Generally speaking we should probably be moving out parts of the logic from InstanceServiceImpl, in order to not bloat it too much.